### PR TITLE
[pytree] add access api

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117771

This PR introduces an API to use KeyPaths to actually access values on pytrees.

Differential Revision: [D52881260](https://our.internmc.facebook.com/intern/diff/D52881260/)

### CODE REVIEW

File: torch/utils/_pytree.py [hunk 1]
```diff
@@ -1476,3 +1497,10 @@ def tree_map_with_path(
 def keystr(kp: KeyPath) -> str:
     """Given a key path, return a pretty-printed representation."""
     return "".join([str(k) for k in kp])
+
+
+def key_get(obj: Any, kp: KeyPath) -> Any:
+    """Given an object and a key path, return the value at the key path."""
+    for k in kp:
+        obj = k.access(obj)
+    return obj
```
Comment (zou3519) [inline torch/utils/_pytree.py:1506]: bikeshedding: the API is called "key_get", but the accessors are called "access". Can we make them  more consistent?

**Review by zou3519 — APPROVED** [review]
LGTM, but I want to bikeshed the naming a bit

File: torch/utils/_pytree.py (hunk 1 above)
Comment (zou3519) [reply torch/utils/_pytree.py:1506]: ideas:
- access_with_key/access ?
- key_get/get ?

Eh, key_get / access is growing on me some more.

Comment (suo) [reply torch/utils/_pytree.py:1506]: lol I hate API design

**Comment by suo** [comment]
> key_get/get

I think this is probably the best if we want them to be the same term, since `key_access` sounds silly. lol

**Review by zou3519 — APPROVED** [review]

**Review by XuehaiPan — APPROVED** [review]

**Pull request closed without merging**